www/js/controller.js

```diff
@@ -692,6 +700,335 @@ define([
+						nodeID: this.currentID,
+						property: "equation",
+						value: inputEquation,
+						correctResult: this._model.given.getEquation(this.currentID),
+						checkResult: "INCORRECT"
+					});
+				}
+				//Check 3 - check if accumulator has a reference to itself as per the Trello card https://trello.com/c/0aqmwqqG
+				if(givenID && this._model.active.getType(this.currentID) === "accumulator" &&
+					givenID === mapID.call(this._model.active, this.currentID)){
+					cancelUpdate = true;
+					directives.push({id: 'equation', attribute: 'status', value: 'incorrect'});
+					directives.push({
+						id: 'crisisAlert',
+						attribute: 'open',
+						value: "The old value of the accumulator is already included in the expression, so you don't have to mention it in the expression.  Only put an expression for the change in the accumulators value."
+					});
+					this.logging.log("solution-step", {
+						type: "self-referencing-accumulator",
+						node: this._model.active.getName(this.currentID),
+						nodeID: this.currentID,
+						property: "equation",
+						value: inputEquation,
+						correctResult: this._model.given.getEquation(this.currentID),
+						checkResult: "INCORRECT"
+					});
+				}
+			}, this);
+
+			if(resetEquation){
+				this._model.active.setEquation(this.currentID, "");
+				directives.push({id: 'equation', attribute: 'value', value: ""});
+			}
+			// changing this as it is essential to update the equation using createExpressionNodes.
+			// otherwise equation with correct nodes not converted to their corresponding id stays in the equation of the model
+			// fix for bug : https://trello.com/c/bVYAQBKT ~ Sachin Grover
+			/*else if(cancelUpdate){
+				//in case we are not calling the pm then we need to save the equation to the model.
+				//this._model.active.setEquation(this.currentID, inputEquation);
+			}*/
+			//if(true || !cancelUpdate){
+			//return parse;
+			//}
+			//return null; */
+
+		},
+		createExpressionNodes: function(parse, ignoreUnknownTest){
+			/*
+			 Create Expression nodes if equation is valid and parsed sucessfully.
+
+			 If the parse succeeds:
+			 * substitute in student id for variable names (when possible),
+			 * save to model,
+			 * update inputs,
+			 * update the associated connections in the graph, and
+			 * send the equation to the PM. **Done**
+			 * Handle the reply from the PM. **Done**
+			 * If the reply contains an update to the equation, the model should also be updated.
+
+			 */
+			 // in topomath parse has noth lhsParse and rhsParse 
+			if(parse){
+				var inputNodesList = [];
+				var cancelUpdate = false;
+				var directives = [];
+
+				var widget = registry.byId(this.controlMap.equation);
+				var inputEquation = widget.get("value");
+
+				var lhsInputs = this.substituteVars(parse.lhsParse,ignoreUnknownTest);
+				var rhsInputs = this.substituteVars(parse.rhsParse,ignoreUnknownTest);
+				inputNodesList = lhsInputs.concat(rhsInputs);
```

Comment:
there is a function in parser which helps us replace the ids with other strings that we want. Check the function convert in equation.js.